modules/memberships/class-jetpack-memberships.php

```diff
@@ -0,0 +1,289 @@
+	 */
+	private function register_init_hook() {
+		add_action( 'init', array( $this, 'init_hook_action' ) );
+	}
+
+	/**
+	 * Actual hooks initializing on init.
+	 */
+	public function init_hook_action() {
+		add_filter( 'rest_api_allowed_post_types', array( $this, 'allow_rest_api_types' ) );
+		add_filter( 'jetpack_sync_post_meta_whitelist', array( $this, 'allow_sync_post_meta' ) );
+		$this->setup_cpts();
+	}
+
+	/**
+	 * Sets up the custom post types for the module.
+	 */
+	private function setup_cpts() {
+		/*
+		 * PLAN data structure.
+		 */
+		$capabilities = array(
+			'edit_post'          => 'edit_posts',
+			'read_post'          => 'read_private_posts',
+			'delete_post'        => 'delete_posts',
+			'edit_posts'         => 'edit_posts',
+			'edit_others_posts'  => 'edit_others_posts',
+			'publish_posts'      => 'publish_posts',
+			'read_private_posts' => 'read_private_posts',
+		);
+		$order_args   = array(
+			'label'               => esc_html__( 'Plan', 'jetpack' ),
+			'description'         => esc_html__( 'Memberships plans', 'jetpack' ),
+			'supports'            => array( 'title', 'custom-fields', 'content' ),
+			'hierarchical'        => false,
+			'public'              => false,
+			'show_ui'             => false,
+			'show_in_menu'        => false,
+			'show_in_admin_bar'   => false,
+			'show_in_nav_menus'   => false,
+			'can_export'          => true,
+			'has_archive'         => false,
+			'exclude_from_search' => true,
+			'publicly_queryable'  => false,
+			'rewrite'             => false,
+			'capabilities'        => $capabilities,
+			'show_in_rest'        => false,
+		);
+		register_post_type( self::$post_type_plan, $order_args );
+	}
+
+	/**
+	 * Allows custom post types to be used by REST API.
+	 *
+	 * @param array $post_types - other post types.
+	 *
+	 * @see hook 'rest_api_allowed_post_types'
+	 * @return array
+	 */
+	public function allow_rest_api_types( $post_types ) {
+		$post_types[] = self::$post_type_plan;
+
+		return $post_types;
+	}
+
+	/**
+	 * Allows custom meta fields to sync.
+	 *
+	 * @param array $post_meta - previously changet post meta.
+	 *
+	 * @return array
+	 */
+	public function allow_sync_post_meta( $post_meta ) {
+		$meta_keys = array_map(
+			array( $this, 'return_meta' ),
+			$this->get_plan_property_mapping()
+		);
+		return array_merge( $post_meta, array_values( $meta_keys ) );
+	}
+
+	/**
+	 * This returns meta attribute of passet array.
+	 * Used for array functions.
+	 *
+	 * @param array $map - stuff.
+	 *
+	 * @return mixed
+	 */
+	public function return_meta( $map ) {
+		return $map['meta'];
+	}
+	/**
+	 * Callback that parses the membership purchase shortcode.
+	 *
+	 * @param array $attrs - attributes in the shortcode. `id` here is the CPT id of the plan.
+	 *
+	 * @return string|void
+	 */
+	public function render_button( $attrs ) {
+		Jetpack_Gutenberg::load_assets_as_required( self::$button_block_name, array( 'thickbox', 'wp-polyfill' ) );
```

Comment:
Note that this function suffers from regression which won't allow `thickbox` to be enqueued, I pointed it out here:

https://github.com/Automattic/jetpack/pull/11591/files#r272845122

We'll need to get that fixed for this PR.

`wp-polyfill` dependency can now be removed; we're including it with every block much like Gutenberg does.

Reply:
In the near-term, I'd like to extend the webpack plugin's API to handle more scripts: https://github.com/Automattic/wp-calypso/issues/32097

For this PR, we should restore the second parameter so we can pass through the `thickbox` dependency 👍 

That will allow this PR to proceed and not rush the webpack plugin API change.

Reply:
https://github.com/Automattic/jetpack/pull/11968